mitxpro/views.py

```diff
@@ -45,6 +47,30 @@ def index(request, **kwargs):  # pylint: disable=unused-argument
     """
     if request.method == "GET":
         return render(request, "index.html", context=get_js_settings_context(request))
+    elif request.method == "POST":
+        response = redirect(request.get_full_path())
+        if (
+            "auth_amount"  # pylint: disable=too-many-boolean-expressions
+            in request.POST
+            and "req_merchant_defined_data2" in request.POST
+            and "req_merchant_defined_data1" in request.POST
+            and "req_reference_number" in request.POST
+            and "req_transaction_uuid" in request.POST
+            and "reason_code" in request.POST
+            and request.POST["reason_code"] == "100"
+        ):
+            payment_dict = {
+                "transaction_id": request.POST["req_transaction_uuid"],
+                "transaction_total": float(request.POST["auth_amount"]),
+                "reference_number": request.POST["req_reference_number"],
+                "product_type": request.POST["req_merchant_defined_data1"],
+                "courseware_id": request.POST["req_merchant_defined_data2"],
+            }
+            # Need to do this in order to avoid escaping the whole thing and the
+            # JSON parser failing on the frontend because of all the escape characters
```

Comment:
I'm heading on vacation so I won't be able to review this, but I think this feels hackish and needs more research. I am curious what the error is here.

Reply:
@noisecapella there is no error as such here, first off the Cybersource payload in the POST body is quite large, setting it in a cookie requires change to the nginx settings (cookie header too large). Most of the data in there isn't needed as per @pdpinch directions so I'm picking off only the info that is of value to us. Secondly I'm just transforming the JSON string so that the quote marks in the key: value format don't get escaped with backslashes (`JSON.parse` fails in that case) and replacing spaces with `%20` so that I can simply use `decodeURIComponent` on the front to get a clean JSON string. That's all.

Reply:
I think @noisecapella is reacting to the use of the string replace instead of purpose-built encoding function. 

I'm also curious about the use of a cookie here. What data specifically are you trying to capture? I'll open another comment thread about that.

Reply:
Yeah, I guess I could encode the whole string as well. I'll update to using that method.

As far as the cookie is concerned, this is the flow for cybersource (from the client perspective):
1. Go to checkout page, select product, click "Pay now"
2. Redirected to Cybersource, fill in info, make payment.
3. Cybersource directs to xPRO with POST on dashboard URL
4. Redirected to same URL with GET (dashboard is displayed finally after purchase)

Now, the Cybersource payment info is only available to xPRO at step 3 (ignoring the server->server instant payment notification), however the response to this step is a redirect to dashboard (step 4) which is displayed in the browser. Step 3 does not load as a template/page response in the client's browser so there is no access to GTM or any other client side libraries yet. The first page that actually loads in the client's browser is at step 4.

We need to carry over the Cybersource payment data from step 3 to step 4, both of which are separate requests and we need to carry the data over to the client's browser, javascript side (GTM is a JS library). The cookie is being used to do that. Once the beacon is sent to GTM, the cookie is deleted (it's a one-off cookie).

Do you have a suggestion for a better workflow?

Reply:
Can we carry that data from step 3 to 4 as a query parameter on the GET request? It sounds like there's a lot of data, but maybe we can just pass a reference to the order id or the basket id and use that to fetch the necessary info.

Reply:
I wouldn't recommend query parameters, I'd rather not expose any such info in the URL that could potentially be tinkered with from the client. Say, a session based value to inject into the page response?

I'm not really concerned about the size of data because we're picking up only a handful of values from the whole POST body. What's your suggestion?

Reply:
A session-based value in the page response sounds OK to me --  is that feasible? 

Cookies aren't really any more secure that URL query parameters, but I understand what you're saying.